homeassistant/helpers/script.py

```diff
@@ -309,173 +222,288 @@ def async_log_exception(self, logger, message_base, exception):
-        self._log("Executing step %s" % self.last_action)
-        event_data = dict(action.get(CONF_EVENT_DATA, {}))
-        if CONF_EVENT_DATA_TEMPLATE in action:
+        self.last_action = self._action.get(CONF_ALIAS, self._action[CONF_EVENT])
+        self._log("Executing step %s", self.last_action)
+        event_data = dict(self._action.get(CONF_EVENT_DATA, {}))
+        if CONF_EVENT_DATA_TEMPLATE in self._action:
             try:
                 event_data.update(
-                    template.render_complex(action[CONF_EVENT_DATA_TEMPLATE], variables)
+                    template.render_complex(
+                        self._action[CONF_EVENT_DATA_TEMPLATE], self._variables
+                    )
                 )
             except exceptions.TemplateError as ex:
-                _LOGGER.error("Error rendering event data template: %s", ex)
+                self._log(
+                    "Error rendering event data template: %s", ex, level=logging.ERROR
+                )
 
-        self.hass.bus.async_fire(action[CONF_EVENT], event_data, context=context)
+        self._hass.bus.async_fire(
+            self._action[CONF_EVENT], event_data, context=self._context
+        )
 
-    async def _async_check_condition(self, action, variables, context):
+    async def _async_condition_step(self):
         """Test if condition is matching."""
-        config_cache_key = frozenset((k, str(v)) for k, v in action.items())
+        config_cache_key = frozenset((k, str(v)) for k, v in self._action.items())
         config = self._config_cache.get(config_cache_key)
         if not config:
-            config = await condition.async_from_config(self.hass, action, False)
+            config = await condition.async_from_config(self._hass, self._action, False)
             self._config_cache[config_cache_key] = config
 
-        self.last_action = action.get(CONF_ALIAS, action[CONF_CONDITION])
-        check = config(self.hass, variables)
-        self._log(f"Test condition {self.last_action}: {check}")
-
+        self.last_action = self._action.get(CONF_ALIAS, self._action[CONF_CONDITION])
+        check = config(self._hass, self._variables)
+        self._log("Test condition %s: %s", self.last_action, check)
         if not check:
             raise _StopScript
 
-    def _async_set_timeout(self, action, variables, context, continue_on_timeout):
-        """Schedule a timeout to abort or continue script."""
-        timeout = action[CONF_TIMEOUT]
-        unsub = None
+    def _log(self, msg, *args, level=logging.INFO):
+        self._script._log(msg, *args, level=level)  # pylint: disable=protected-access
+
+    def _raise(self, msg, *args, exception=None):
+        # pylint: disable=protected-access
+        self._script._raise(msg, *args, exception=exception)
+
+
+class _ScriptRun(_ScriptRunBase):
+    """Manage asynchronous run of Script sequence."""
+
+    def __init__(
```

Comment:
You can also drop the `__init__` and just add this
```suggestion
    task: Optional[asyncio.Task] = None
```

Reply:
I suppose. I've never used that style of coding. But, then again, I'm still relatively new to Python. (I coded in C++ much more extensively.) Does it have any impact on performance, or is it just a style issue?

Reply:
Ah, ok. I read up on class attributes vs instance attributes and this makes sense. Will do.